File: http-servers/typescript/bun-honojs/src/database/mongodb.ts
import { MongoClient, ObjectId } from "mongodb";
import type { CreateUser, UpdateUser, User, UserRepository } from "./repo";

type UserDocument = {
  _id: ObjectId;
  name: string;
  email: string;
};

export class MongoUserRepository implements UserRepository {
  private client: MongoClient;
  private dbName: string;
  private ready: Promise<void> | null = null;

  constructor(connectionString: string, dbName: string) {
    this.client = new MongoClient(connectionString);
    this.dbName = dbName;
  }

  private async connect(): Promise<void> {
    if (!this.ready) {
      this.ready = this.client.connect().then(() => undefined);
    }
    await this.ready;
  }

  private collection() {
    return this.client.db(this.dbName).collection<UserDocument>("users");
  }

  async create(data: CreateUser): Promise<User> {
    await this.connect();
    const id = new ObjectId();
    await this.collection().insertOne({ _id: id, name: data.name, email: data.email });
    return { id: id.toString(), name: data.name, email: data.email };
  }

  async findById(id: string): Promise<User | null> {
    await this.connect();
    let objectId: ObjectId;
    try {
      objectId = new ObjectId(id);
    } catch {
      return null;
    }

    const user = await this.collection().findOne({ _id: objectId });
    if (!user) {
      return null;
    }

    return { id: user._id.toString(), name: user.name, email: user.email };
  }

  async update(id: string, data: UpdateUser): Promise<User | null> {
    await this.connect();
    let objectId: ObjectId;
    try {
      objectId = new ObjectId(id);
    } catch {
      return null;
    }

    const user = await this.collection().findOneAndUpdate(
      { _id: objectId },
      { $set: { name: data.name, email: data.email } },
      { returnDocument: "after" },
    );

    if (!user) {
      return null;
    }

    return { id: user._id.toString(), name: user.name, email: user.email };
  }

  async delete(id: string): Promise<boolean> {
    await this.connect();
    let objectId: ObjectId;
    try {
      objectId = new ObjectId(id);
    } catch {
      return false;
    }

    const result = await this.collection().deleteOne({ _id: objectId });
    return result.deletedCount > 0;
  }

  async deleteAll(): Promise<void> {
    await this.connect();
    await this.collection().deleteMany({});
  }

  async healthCheck(): Promise<boolean> {
    try {
      await this.connect();
      await this.client.db(this.dbName).command({ ping: 1 });
      return true;
    } catch {
      return false;
    }
  }

  async disconnect(): Promise<void> {
    await this.client.close();
  }
}

File: http-servers/python/fastapi/src/database/mongodb.py
from __future__ import annotations

from bson import ObjectId
from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorCollection
from pymongo import ReturnDocument

from src.database.types import CreateUser, UpdateUser, User, build_user


class MongoUserRepository:
    def __init__(self, connection_string: str, db_name: str):
        self._url = connection_string
        self._db_name = db_name
        self._client: AsyncIOMotorClient | None = None

    def _collection(self) -> AsyncIOMotorCollection:
        if self._client is None:
            self._client = AsyncIOMotorClient(self._url)
        return self._client[self._db_name]["users"]

    def _parse_object_id(self, id: str) -> ObjectId | None:
        try:
            return ObjectId(id)
        except Exception:
            return None

    def _to_user(self, doc: dict) -> User:
        return User(
            id=str(doc["_id"]),
            name=doc["name"],
            email=doc["email"],
            favoriteNumber=doc.get("favoriteNumber"),
        )

    async def create(self, data: CreateUser) -> User:
        id = ObjectId()
        doc = {"_id": id, "name": data.name, "email": data.email}
        if data.favoriteNumber is not None:
            doc["favoriteNumber"] = data.favoriteNumber

        await self._collection().insert_one(doc)
        return build_user(str(id), data)

    async def find_by_id(self, id: str) -> User | None:
        oid = self._parse_object_id(id)
        if oid is None:
            return None

        doc = await self._collection().find_one({"_id": oid})
        return self._to_user(doc) if doc else None

    async def update(self, id: str, data: UpdateUser) -> User | None:
        oid = self._parse_object_id(id)
        if oid is None:
            return None

        update_fields = {}
        if data.name is not None:
            update_fields["name"] = data.name
        if data.email is not None:
            update_fields["email"] = data.email
        if data.favoriteNumber is not None:
            update_fields["favoriteNumber"] = data.favoriteNumber

        if not update_fields:
            return await self.find_by_id(id)

        doc = await self._collection().find_one_and_update(
            {"_id": oid}, {"$set": update_fields}, return_document=ReturnDocument.AFTER
        )
        return self._to_user(doc) if doc else None

    async def delete(self, id: str) -> bool:
        oid = self._parse_object_id(id)
        if oid is None:
            return False

        result = await self._collection().delete_one({"_id": oid})
        return result.deleted_count > 0

    async def delete_all(self) -> None:
        await self._collection().delete_many({})

    async def health_check(self) -> bool:
        try:
            if self._client is None:
                self._client = AsyncIOMotorClient(self._url)
            await self._client.admin.command("ping")
            return True
        except Exception:
            return False

    async def disconnect(self) -> None:
        if self._client is not None:
            self._client.close()
            self._client = None
